gui/msgwin.c

```diff
@@ -0,0 +1,249 @@
+/**
+ * @file
+ * Message Window
+ *
+ * @authors
+ * Copyright (C) 2021 Richard Russon <rich@flatcap.org>
+ *
+ * @copyright
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, either version 2 of the License, or (at your option) any later
+ * version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @page gui_msgwin Message Window
+ *
+ * Message Window for displaying messages and text entry.
+ */
+
+#include "config.h"
+#include <stddef.h>
+#include "mutt/lib.h"
+#include "core/lib.h"
+#include "color.h"
+#include "curs_lib.h"
+#include "mutt_curses.h"
+#include "mutt_window.h"
+
+struct MuttWindow *MessageWindow = NULL; ///< Message Window, ":set", etc
+
+/**
+ * struct MsgWinPrivateData - Private data for the Message Window
+ */
+struct MsgWinPrivateData
+{
+  enum ColorId color; ///< Colour for the text, e.g. #MT_COLOR_MESSAGE
+  char *text;         ///< Cached display string
+};
+
+/**
+ * msgwin_recalc - Recalculate the display of the Message Window
+ * @param win Message Window
+ */
+static int msgwin_recalc(struct MuttWindow *win)
+{
+  if (!win)
+    return -1;
+
+  struct MuttWindow *win_focus = window_get_focus();
+  if (!win_focus)
+    return 0;
+
+  // If the MessageWindow is focussed, then someone else is actively using it.
+  if (win_focus == win)
+    return 0;
+
+  // If nobody is using it, we'll take charge
+  win->actions |= WA_REPAINT;
+  mutt_debug(LL_DEBUG5, "recalc done, request WA_REPAINT\n");
+  return 0;
+}
+
+/**
+ * msgwin_repaint - Redraw the Message Window
+ * @param win Message Window
+ */
+static int msgwin_repaint(struct MuttWindow *win)
+{
+  if (!mutt_window_is_visible(win))
+    return 0;
+
+  struct MuttWindow *win_focus = window_get_focus();
+  if (!win_focus)
+    return 0;
+
+  // If the MessageWindow is focussed, then someone else is actively using it.
+  if (win_focus == win)
+    return 0;
```

Comment:
Since this block of code appears twice and neither occurrences uses `win_focus` outside of the checks, it makes sense to move it to a well-named function so easier to focus on the important logic:

```C
static bool mutt_window_is_focused(MuttWindow *win)
{
   struct MuttWindow *win_focus = window_get_focus();
    if (!win_focus)
      return false;

    // If the MessageWindow is focussed, then someone else is actively using it.
    if (win_focus == win)
      return false;
        
    return true;
}

static int msgwin_repaint(struct MuttWindow *win)
{
  if (!mutt_window_is_visible(win) || !mutt_window_is_focused(win))
    return 0;

  struct MsgWinPrivateData *priv = win->wdata;

  mutt_window_move(win, 0, 0);

  mutt_curses_set_color(priv->color);
  mutt_window_move(win, 0, 0);
  mutt_paddstr(win, win->state.cols, priv->text);
  mutt_curses_set_color(MT_COLOR_NORMAL);

  mutt_debug(LL_DEBUG5, "repaint done\n");
  return 0;
}
```

Reply:
Good call.  Done.

My defensive style of programming means I've put in a huge number of checks.
Soon, I may be able to remove a lot of them.

For `recalc()` and `repaint()`, I'll guarantee `win` to be non-`NULL` as a contract of the function.

In the future, the Window functions won't get called if the Window isn't visible.

Finally, many of the observers watch `NeoMutt->notify` which leads to a LOT of notifications that aren't of interest.  By changing the observed object to a **single** object, more checks can be removed.